WIP: Feat/args

TomFrost opened the pull request:
Ran out of time so I'll review and repair tonight -- the run script hangs and I haven't had time to debug.

TomFrost committed:
fix(entrypoint): Don't ignore docker commands

TomFrost committed:
feat(run): Lots of options

TomFrost committed:
Merge branch 'master' into feat/args

ksafranski commented on run.sh:
```diff
@@ -1,31 +1,77 @@
+  echo ""
+  echo "  -i  Launch the container interactively and open the shell prompt."
+  echo "      Otherwise, the container will run in daemon mode."
+  echo "  -m  Mount the .ssh and .aws directories from the host as volumes in"
```
I like this, but I wonder if we're being too restrictive with these specifics. What if `-m` accepted a comma-delimited set of paths?

TomFrost replied:
Good call. I was erring on the side of simpler arguments (since, at a certain point, you might as well just be typing a docker build command yourself), but I think expressive-as-possible makes more sense since we'll probably want to alias our config for this on a per-env basis anyway.

To simplify, though, rather than do string processing in bash, I'll just allow for multiple `-m` arguments to be passed, relative to $HOME. So the current functionality would be replicated with `-m .ssh -m .aws`.

Fluidbyte commented:
Two needs (just so we don't forget):

- `-a <PORT>`: Set the application port (`-p <PORT>:3000` in the script)
- `-o <PORT(S),RANGE(S)>`: Opens ports (i.e. `-o 5555,8000-8100` >> `-p 5555:5555 -p 8000-8100:8000-8100`)

TomFrost committed:
feat(Dockerfile): sshd, usergroup, use command

TomFrost committed:
feat(entrypoint): Start ssh

TomFrost committed:
feat(run.sh): All the things.

TomFrost commented:
So many things.
- Added `-a` and `-o` per above
- Made parallel volume mappings with `-m` explicit per above
- Consolidated some of the user stuff and created the missing user group
- Added sshd. Now you can ssh into the box while it's running (see readme)
- Turned the entrypoint into a real entrypoint -- commands work now. As such, code-server is now a command, but you can also start the container directly into a shell without running code-server if you want (not built into run.sh though -- I don't think we need it there)
- Made the home folder mount optional. This makes local dev on a mac _scream_ compared to working over a file share. Testing svc-marketplace takes about 60% of the time it used to now.
- Removed the --rm because, if you're not working in a volume, this could unintentionally delete all your work. Now that it's out, you can start and stop the container at-will without losing anything. Maybe we can surface a flag in run.sh if it's useful to have for volume-based workflows.
- The username now defaults to your machine username, so there's no real reason to specify -u anymore.
- Updated the readme. I removed the stuff about chmodding run.sh -- git stores the mode of each file, so the `+x` was committed and pulls down when you clone the repo now. <3 git.

My `run.sh` command is now:

```shell
./run.sh -s zsh -o 8000-8100 -m .ssh -m .aws -m .dotfiles
```

:boom:

Random stuff in my head:
- It might make sense, for people not volumizing their workspace, to offer an nfs or smb share or something if it's lightweight. There are some quirks to code-server that I'd love to bypass by mounting the container's filesystem back on the host and using vscode instead.
- Might be just because I'm virtualizing on mac, but some docker operations are REALLY slow and occasionally buggy. I had a _hell_ of a time pulling `draynow/node-base:functions`. Granted, that one's HUGE, but I was getting errors right and left and when it was finally successful it took over 10 minutes.

Ok I think that's it. I should sleep. But this was a ton of fun.

Fluidbyte commented:
This looks great. 👍 on the SSH service, that was one thing I was considering for the future; I like the idea of having an environment similar to a completely containerized VPS with the benefit of having VSCode available.

I noticed networking was slow as hell on a Mac, and we already knew the whole FS issue with mounts. I didn't have any issues (across multiple runs) with it _not_ working, just took forever. Not really sure what direction to go.

I'll dig further on the PR today and run through testing/QA locally and on my remote box.
